### PR TITLE
Feat/singlepass dylib

### PR DESCRIPTION
**Pull request by ptitSeb**
# Description
Enable Dylib+singlepass on Linux and macOS. Works fine, but larger wasm file may still fail to link on macOS/Aarch64.
For #2736

### CODE REVIEW

File: lib/engine-dylib/src/artifact.rs
Comment (Amanieu) [inline lib/engine-dylib/src/artifact.rs:387]: typo?

File: lib/engine-dylib/src/artifact.rs
Comment (Amanieu) [inline lib/engine-dylib/src/artifact.rs:405]: We actually want to remove both of these options. They currently tell the linker to allow text relocations which we *don't* want for PIC code.

**Comment by syrusakbary** [comment]
I don't have enough context on this PR to approve myself. So we'll wait for @Amanieu review!

**Comment by Amanieu** [comment]
I tested this branch and it seems to generate a .so file with zero runtime relocations, which is good. However I don't see why the `no-text` flag is still needed. Additionally, this PR has a lot of unnecessary changes like:
- `emit_ldr_label` and `emit_nop` that are never used.
- Leftover change to traphandlers.c
- It still includes the linker flags that we don't want. This should never trigger if we are indeed emitting code with zero relocations.

**Comment by ptitSeb** [comment]
> * It still includes the linker flags that we don't want. This should never trigger if we are indeed emitting code with zero relocations.

The cross compile test is failling. It compile using dylib-cranelift targetting x86_64-linux-musl and fail with a relocation of type `R_X86_64_64` for `wasmer_vm_probestack` in text. It only happens for musl. It seems to fails if I actually cross compile: if I cange the example to build to x86_64-linux-gnu on an x86_64 linux box, it works, but the same change on an arm64 box the sample fails. It seems to only happens when `lld` is used. `ld`/`gold` don't have the issue.

**Comment by Amanieu** [comment]
Can you dump the object files that are passed to the linker in both the success and failing cases? Also the arguments passed to the linker. The `wasmer_vm_probestack` symbol should be defined in the same object file, in a separate section that is added by engine-dylib.

**Comment by ptitSeb** [comment]
Yes, I was thinking on the same thing, check the difference, if any, on both object files.

**Comment by Amanieu** [comment]
Actually the `R_X86_64_64` relocation seems strange. That's an absolute 64-bit relocation. I would have expected a `R_X86_64_PC32` instead which is used for the offset in a call instruction.

I think that I know what the problem is, can you try applying this commit from #2807: https://github.com/wasmerio/wasmer/pull/2807/commits/657bbae518a649ecc98b06571a78f30fdb1fe023

You might also need to apply this fix: https://github.com/wasmerio/wasmer/commit/a2809145a1af7efa6cffc565f2144475e8af5305

**Comment by ptitSeb** [comment]
Yes, it works with both commit.

**Comment by ptitSeb** [comment]
So I'll wait until the Stack/Corosensei branch is merged...

File: lib/object/src/module.rs
Comment (Amanieu) [inline lib/object/src/module.rs:314]: Are these still needed?

**Review by Amanieu — APPROVED** [review]

File: lib/object/src/module.rs
Comment (ptitSeb) [reply lib/object/src/module.rs:314]: Probably not

Comment (Amanieu) [reply lib/object/src/module.rs:314]: Remove them then it's good to r+.

**Review by Amanieu — APPROVED** [review]

**Pull request merged**
